[quant] Input-weight equalization - branch support

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62366

Summary: In the case of models with branches, we are unable to equalize the branching part in the graph.

For example, given this graph:
```
     conv2 
    /     \
x -> conv1 -> add
```

After prepare, we will ignore the branched layers (conv1 and conv2) and will not insert the equalization observers. A warning message will also be printed with the layers that are unable to be equalized.
```
                        conv2 -> out_quant_obs2
                       /                       \
x -> input_quant_obs -> conv1 -> out_quant_obs1 -> add 
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29982585](https://our.internmc.facebook.com/intern/diff/D29982585)

angelayi committed:
[quant] Input-weight equalization - branch support

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-weight equalization - branch support"


Summary: In the case of models with branches, we are unable to equalize the branching part in the graph.

For example, given this graph:
```
     conv2 
    /     \
x -> conv1 -> add
```

After prepare, we will ignore the branched layers (conv1 and conv2) and will not insert the equalization observers. A warning message will also be printed with the layers that are unable to be equalized.
```
                        conv2 -> out_quant_obs2
                       /                       \
x -> input_quant_obs -> conv1 -> out_quant_obs1 -> add 
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-weight equalization - branch support"


Summary: In the case of models with branches, we are unable to equalize the branching part in the graph.

For example, given this graph:
```
     conv2 
    /     \
x -> conv1 -> add
```

After prepare, we will ignore the branched layers (conv1 and conv2) and will not insert the equalization observers. A warning message will also be printed with the layers that are unable to be equalized.
```
                        conv2 -> out_quant_obs2
                       /                       \
x -> input_quant_obs -> conv1 -> out_quant_obs1 -> add 
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29982585).

supriyar commented on torch/quantization/fx/prepare.py:
```diff
@@ -940,6 +948,24 @@ def insert_observers_for_model(
+                    # conv1 and conv2 are both in branches because its previous
+                    # node (x) has 2 users (conv1 and conv2). These layers will
+                    # currently not get equalized.
+                    is_branch = (
```
Can we move this check inside `maybe_insert_input_equalization_observers_for_node`?

angelayi replied:
I wanted to keep the check inside `maybe_insert_input_equalization_observers_for_node`, but the reason why I had it before was because the previous `maybe_insert_input_observers_for_node` modifies the graph by inserting quantization observers.

For example:
```
     conv2 
    /     \
x -> conv1 -> add
```

Looking at conv1, insert_input_observers() will insert a quantization observer and the graph will be modified as follows:
```
             conv2 
    /                       \
x -> inp_quant_obs -> conv1 -> add
```
So then when we call insert_equalization_observers(), we would need to check 2 nodes before conv1 to see if it has 2 users and is in a branch.

But when we are looking at conv2, insert_input_observers() will modify the graph to be the following:
```
                       conv2 
                      /     \
x -> inp_quant_obs -> conv1 -> add
```
In this case we could check the 1 node before conv2 and see if it has 2 users to see if conv2 is in a branch.

So, the loops would need two different casings, which is why I decided to add the check for whether the node is in a branch before the insert_input_observer() function. Does that make sense?

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -250,23 +253,29 @@ def test_input_weight_equalization_prepare(self):
+                 (ConvReluConvModel, default_qconfig_dict, two_nn_layer_node_occurrence),
+                 (FunctionalConvReluModel, default_qconfig_dict, single_F_layer_node_occurrence),
+                 (FunctionalConvReluConvModel, default_qconfig_dict, two_F_layer_node_occurrence),
+                 (ModelMultipleOpsNoAvgPool, default_qconfig_dict, multiple_ops_node_occurrence)]
```
is ModelMultipleOpsNoAvgPool testing the branch case? Maybe we can have a more directed test case for this that checks to make sure the warning was raised since `test_input_weight_equalization_prepare` seems to be getting very complex.

angelayi committed:
Update on "[quant] Input-weight equalization - branch support"


Summary: In the case of models with branches, we are unable to equalize the branching part in the graph.

For example, given this graph:
```
     conv2 
    /     \
x -> conv1 -> add
```

After prepare, we will ignore the branched layers (conv1 and conv2) and will not insert the equalization observers. A warning message will also be printed with the layers that are unable to be equalized.
```
                        conv2 -> out_quant_obs2
                       /                       \
x -> input_quant_obs -> conv1 -> out_quant_obs1 -> add 
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29982585](https://our.internmc.facebook.com/intern/diff/D29982585)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/prepare.py:
```diff
@@ -940,6 +949,34 @@ def insert_observers_for_model(
+                            )
+                            if user != node and is_user_quantized:
+                                is_quantized_branch = True
+                                print(is_quantized_branch, user)
+                                break
```
Is the print and break here intended?

angelayi replied:
Oh no, sorry let me remove that.

angelayi committed:
Update on "[quant] Input-weight equalization - branch support"


Summary: In the case of models with branches, we are unable to equalize the branching part in the graph.

For example, given this graph:
```
     conv2 
    /     \
x -> conv1 -> add
```

After prepare, we will ignore the branched layers (conv1 and conv2) and will not insert the equalization observers. A warning message will also be printed with the layers that are unable to be equalized.
```
                        conv2 -> out_quant_obs2
                       /                       \
x -> input_quant_obs -> conv1 -> out_quant_obs1 -> add 
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29982585](https://our.internmc.facebook.com/intern/diff/D29982585)

[ghstack-poisoned]

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -267,6 +270,43 @@ def test_input_weight_equalization_prepare(self):
+        }
+
+        tests = [
+            (ModelMultipleOpsNoAvgPool, default_qconfig_dict, multiple_ops_node_occurrence),
+            (AddBranchingModel, specific_qconfig_dict, add_branching_node_occurrence),
```
Trying to understand these tests a little better. Is ModelMultipleOpsNoAvgPool supposed to test the case where we don't equalize across branches (print warning) since all nodes are being quantized? And is AddBranchingModel supposed to test that we still equalize in a branch if the other branch is in FP32?
Maybe making the names a bit more descriptive and adding some comments might help someone trying to read this code.

angelayi replied:
Yes, I will add print statements to ensure that these are better understood!

angelayi committed:
Update on "[quant] Input-weight equalization - branch support"


Summary: In the case of models with branches, we are unable to equalize the branching part in the graph.

For example, given this graph:
```
     conv2 
    /     \
x -> conv1 -> add
```

After prepare, we will ignore the branched layers (conv1 and conv2) and will not insert the equalization observers. A warning message will also be printed with the layers that are unable to be equalized.
```
                        conv2 -> out_quant_obs2
                       /                       \
x -> input_quant_obs -> conv1 -> out_quant_obs1 -> add 
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29982585](https://our.internmc.facebook.com/intern/diff/D29982585)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-weight equalization - branch support"


Summary: In the case of models with branches, we are unable to equalize the branching part in the graph.

For example, given this graph:
```
     conv2 
    /     \
x -> conv1 -> add
```

After prepare, we will ignore the branched layers (conv1 and conv2) and will not insert the equalization observers. A warning message will also be printed with the layers that are unable to be equalized.
```
                        conv2 -> out_quant_obs2
                       /                       \
x -> input_quant_obs -> conv1 -> out_quant_obs1 -> add 
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29982585](https://our.internmc.facebook.com/intern/diff/D29982585)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29982585).

angelayi committed:
Update on "[quant] Input-weight equalization - branch support"


Summary: In the case of models with branches, we are unable to equalize the branching part in the graph.

For example, given this graph:
```
     conv2 
    /     \
x -> conv1 -> add
```

After prepare, we will ignore the branched layers (conv1 and conv2) and will not insert the equalization observers. A warning message will also be printed with the layers that are unable to be equalized.
```
                        conv2 -> out_quant_obs2
                       /                       \
x -> input_quant_obs -> conv1 -> out_quant_obs1 -> add 
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29982585](https://our.internmc.facebook.com/intern/diff/D29982585)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29982585).

angelayi committed:
Update on "[quant] Input-weight equalization - branch support"


Summary: In the case of models with branches, we are unable to equalize the branching part in the graph.

For example, given this graph:
```
     conv2 
    /     \
x -> conv1 -> add
```

After prepare, we will ignore the branched layers (conv1 and conv2) and will not insert the equalization observers. A warning message will also be printed with the layers that are unable to be equalized.
```
                        conv2 -> out_quant_obs2
                       /                       \
x -> input_quant_obs -> conv1 -> out_quant_obs1 -> add 
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_prepare`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29982585](https://our.internmc.facebook.com/intern/diff/D29982585)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29982585).
